File: packages/sardis-stagehand/src/tools.ts
/**
 * Sardis payment tools for Stagehand browser agents.
 */
import { z } from 'zod';

interface SardisToolsConfig {
  apiKey?: string;
  walletId?: string;
  chain?: string;
  token?: string;
}

export function createSardisTools(config: SardisToolsConfig = {}) {
  const apiKey = config.apiKey || process.env.SARDIS_API_KEY || '';
  const defaultWalletId = config.walletId || process.env.SARDIS_WALLET_ID || '';
  const defaultChain = config.chain || 'base';
  const defaultToken = config.token || 'USDC';

  // Lazy client initialization to avoid requiring @sardis/sdk at import time
  let _client: any = null;
  const getClient = () => {
    if (!_client) {
      // Dynamic import pattern - works with or without @sardis/sdk
      try {
        const { SardisClient } = require('@sardis/sdk');
        _client = new SardisClient({ apiKey });
      } catch {
        throw new Error('@sardis/sdk is required: npm install @sardis/sdk');
      }
    }
    return _client;
  };

  return {
    sardis_pay: {
      name: 'sardis_pay',
      description: 'Execute a policy-controlled payment from the agent\'s Sardis wallet',
      schema: z.object({
        amount: z.number().describe('Payment amount in USD'),
        merchant: z.string().describe('Merchant or recipient identifier'),
        purpose: z.string().optional().describe('Reason for payment'),
        walletId: z.string().optional().describe('Wallet ID override'),
      }),
      execute: async ({ amount, merchant, purpose, walletId }: {
        amount: number;
        merchant: string;
        purpose?: string;
        walletId?: string;
      }) => {
        const client = getClient();
        const wid = walletId || defaultWalletId;
        if (!wid) return { success: false, error: 'No wallet ID configured' };

        try {
          const result = await client.wallets.transfer(wid, {
            destination: merchant,
            amount: amount.toString(),
            token: defaultToken,
            chain: defaultChain,
            memo: purpose,
          });
          return {
            success: true,
            status: 'APPROVED',
            transactionHash: result.tx_hash,
            message: `Payment of $${amount} to ${merchant} completed`,
          };
        } catch (error: unknown) {
          const msg = error instanceof Error ? error.message : 'Payment failed';
          return { success: false, status: 'BLOCKED', error: msg };
        }
      },
    },

    sardis_check_balance: {
      name: 'sardis_check_balance',
      description: 'Check the current wallet balance before making a purchase',
      schema: z.object({
        walletId: z.string().optional().describe('Wallet ID override'),
        token: z.string().optional().describe('Token to check'),
      }),
      execute: async ({ walletId, token }: { walletId?: string; token?: string }) => {
        const client = getClient();
        const wid = walletId || defaultWalletId;
        if (!wid) return { success: false, error: 'No wallet ID configured' };

        try {
          const balance = await client.wallets.getBalance(wid, defaultChain, token || defaultToken);
          return {
            success: true,
            balance: balance.balance,
            token: balance.token,
            chain: balance.chain,
          };
        } catch (error: unknown) {
          return { success: false, error: error instanceof Error ? error.message : 'Failed' };
        }
      },
    },

    sardis_check_policy: {
      name: 'sardis_check_policy',
      description: 'Check if a purchase would be allowed by spending policy',
      schema: z.object({
        amount: z.number().describe('Amount to check'),
        merchant: z.string().describe('Merchant to check'),
      }),
      execute: async ({ amount, merchant }: { amount: number; merchant: string }) => {
        const client = getClient();
        const wid = defaultWalletId;
        if (!wid) return { success: false, error: 'No wallet ID configured' };

        try {
          const wallet = await client.wallets.get(wid);
          const limitPerTx = parseFloat(wallet.limit_per_tx || '0');
          return {
            allowed: amount <= limitPerTx,
            message: amount <= limitPerTx
              ? `$${amount} to ${merchant} would be allowed`
              : `$${amount} exceeds per-tx limit of $${limitPerTx}`,
          };
        } catch (error: unknown) {
          return { success: false, error: error instanceof Error ? error.message : 'Failed' };
        }
      },
    },
  };
}

File: packages/sardis-browser-use/sardis_browser_use/tools.py
"""Sardis payment tools for Browser Use agents."""
from __future__ import annotations

import os

from sardis import SardisClient


def _get_client(api_key: str | None = None, wallet_id: str | None = None):
    key = api_key or os.getenv("SARDIS_API_KEY")
    wid = wallet_id or os.getenv("SARDIS_WALLET_ID")
    client = SardisClient(api_key=key)
    return client, wid


def register_sardis_actions(controller, *, api_key: str | None = None, wallet_id: str | None = None):
    """Register all Sardis payment actions on a Browser Use controller.

    Args:
        controller: Browser Use Controller instance
        api_key: Sardis API key (or set SARDIS_API_KEY env var)
        wallet_id: Default wallet ID (or set SARDIS_WALLET_ID env var)
    """
    client, default_wallet_id = _get_client(api_key, wallet_id)

    @controller.action("Pay for a product or service using Sardis wallet with spending policy controls")
    async def sardis_pay(amount: float, merchant: str, purpose: str = "Purchase") -> str:
        wid = default_wallet_id
        if not wid:
            return "Error: No wallet ID configured. Set SARDIS_WALLET_ID or pass wallet_id."
        result = client.payments.send(wid, to=merchant, amount=amount, purpose=purpose)
        if result.success:
            return f"APPROVED: ${amount} to {merchant} (tx: {result.tx_id})"
        return f"BLOCKED by policy: {result.message}"

    @controller.action("Check wallet balance before making a purchase")
    async def sardis_balance(token: str = "USDC") -> str:
        wid = default_wallet_id
        if not wid:
            return "Error: No wallet ID configured."
        balance = client.wallets.get_balance(wid, token=token)
        return f"Balance: ${balance.balance} {token} | Remaining limit: ${balance.remaining}"

    @controller.action("Check if a purchase would be allowed by Sardis spending policy")
    async def sardis_check_policy(amount: float, merchant: str) -> str:
        wid = default_wallet_id
        if not wid:
            return "Error: No wallet ID configured."
        balance = client.wallets.get_balance(wid)
        if amount > balance.remaining:
            return f"WOULD BE BLOCKED: ${amount} exceeds remaining limit ${balance.remaining}"
        if amount > balance.balance:
            return f"WOULD BE BLOCKED: ${amount} exceeds balance ${balance.balance}"
        return f"WOULD BE ALLOWED: ${amount} to {merchant} (balance: ${balance.balance}, remaining: ${balance.remaining})"

    return [sardis_pay, sardis_balance, sardis_check_policy]
